quic-forward-proxy/src/outbound/tpu_connection_manager.rs

```diff
@@ -0,0 +1,283 @@
+        register_int_gauge!(opts!("literpc_quic_tasks", "Number of connections to keep asked by tpu service")).unwrap();
+}
+
+
+const MAXIMUM_TRANSACTIONS_IN_QUEUE: usize = 16_384;
+const SHUTDOWN_AGENT_THRESHOLD: Duration = Duration::from_millis(2500);
+
+pub type WireTransaction = Vec<u8>;
+
+#[derive(Debug, Clone)]
+pub enum BroadcastMessage {
+    Transaction(WireTransaction),
+    Shutdown(Pubkey),
+}
+
+struct ActiveConnection {
+    endpoints: RotatingQueue<Endpoint>,
+    identity: Pubkey,
+    tpu_address: SocketAddr,
+    exit_signal: Arc<AtomicBool>,
+    connection_parameters: QuicConnectionParameters,
+    last_used: Arc<AtomicTiming>,
+}
+
+impl ActiveConnection {
+    pub fn new(
+        endpoints: RotatingQueue<Endpoint>,
+        tpu_address: SocketAddr,
+        identity: Pubkey,
+        connection_parameters: QuicConnectionParameters,
+    ) -> Self {
+        Self {
+            endpoints,
+            tpu_address,
+            identity,
+            exit_signal: Arc::new(AtomicBool::new(false)),
+            connection_parameters,
+            last_used: Arc::new(AtomicTiming::default()),
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    async fn listen(
+        connection_pool: QuicConnectionPool,
+        broadcast_receiver: Receiver<BroadcastMessage>,
+        exit_signal: Arc<AtomicBool>,
+        identity: Pubkey,
+        addr: SocketAddr,
+        last_used: Arc<AtomicTiming>,
+    ) {
+        NB_QUIC_ACTIVE_CONNECTIONS.inc();
+        let mut broadcast_receiver = broadcast_receiver;
+
+        'all_message: loop {
+
+            tokio::select! {
+                broadcast_message = broadcast_receiver.recv() => {
+
+                    match broadcast_message {
+                        Ok(BroadcastMessage::Transaction(tx_raw)) => {
+                            last_used.update();
+                            let PooledConnection {
+                                connection,
+                                permit
+                            } = match connection_pool.get_pooled_connection().await {
+                                Ok(connection_pool) => connection_pool,
+                                Err(e) => {
+                                    error!("error getting pooled connection {e:?}");
+                                    break 'all_message;
+                                },
+                            };
+
+                            tokio::spawn(async move {
+                                // permit will be used to send all the transaction and then destroyed
+                                let _permit = permit;
+                                NB_QUIC_TASKS.inc();
+                                connection.send_transaction(tx_raw).await;
+                                NB_QUIC_TASKS.dec();
+                            });
+
+                        },
+                        Ok(BroadcastMessage::Shutdown(tpu_identity)) => {
```

Comment:
wdyt? sending _shutdown_ in line with transaction payload in same channel will remove race condition between transactions remaining in queue while connection thread is shutting down

Reply:
Ok i think, removes one extra channel.